python/ray/serve/backend_state.py

```diff
@@ -78,35 +113,139 @@ def get_replica_handles(
     def get_backend(self, backend_tag: BackendTag) -> Optional[BackendInfo]:
         return self.backends.get(backend_tag)
 
+    def num_pending_goals(self) -> int:
+        return len(self.pending_goals)
+
+    async def wait_for_goal(self, goal_id: GoalId) -> None:
+        start = time.time()
+        if goal_id not in self.pending_goals:
+            logger.debug(f"Goal {goal_id} not found")
+            return True
+        event = self.pending_goals[goal_id]
+        await event.wait()
+        logger.debug(
+            f"Waiting for goal {goal_id} took {time.time() - start} seconds")
+
+    def _complete_goal(self, goal_id: GoalId) -> None:
+        logger.debug(f"Completing goal {goal_id}")
+        event = self.pending_goals.pop(goal_id, None)
+        if event:
+            event.set()
+
+    def _create_goal(self, goal_id: Optional[GoalId] = None) -> GoalId:
+        if goal_id is None:
+            goal_id = uuid4()
+        event = asyncio.Event()
+        self.pending_goals[goal_id] = event
+        return goal_id
```

Comment:
All of this logic could/should be moved into a separate class that could in the future be shared between `State` classes, but wanted to minimize changes